### DISCUSSION TITLE
Initialisation of store @main

### DISCUSSION THREAD

**Discussion by bielikb** [pointfreeco/swift-composable-architecture#410] [Q&A] [answered]
What's the preferred way to declare the store at the root level of the SwiftUI app -> @main struct? 

As this is the entry point the app communicates with, I no longer could rely on dependency injection, thus I statically initialized the store, that is declared as the private member of the app.

```swift
@main
struct ScrumdingerApp: App {
    ...
    
    private let store: Store<AppState, AppAction> = {
        return Store(initialState: AppState(scrumsState: ScrumsViewState()),
                     reducer: reducer,
                     environment: appEnvironment)
    }()
    
    ...
}
``` 
My current implementation can be seen [here](https://github.com/bielikb/scrumdinger/blob/aff2606ec16bcc29105a3230d98815334213efde/Scrumdinger/ScrumdingerApp.swift#L79).

Is this the right way to do so, though?
Is there something that can be done towards gaining the testability of this main store? 
The Scrumdinger app itself declares the actions to save and load the scrums from the persistent store at the root level of the SwiftUI app.

Thanks!

Best,
Boris

**Comment by stephencelis** [comment, marked as answer]
> Is this the right way to do so, though?

This is where we'd do it, for sure! Our demo apps are written to support iOS 13, so they don't demonstrate this, but we'll probably upgrade them sometime soon, and definitely by the next version of iOS.

> Is there something that can be done towards gaining the testability of this main store?

Can you clarify what you mean by testability here and what kind of testability you would have _outside_ TCA? Generally TCA testing is done via unit tests and the `TestStore` class, or snapshot tests will most likely focus more on a feature's `View` and not an application's entire `App`. I'm not aware of any guidance from Apple for testing `App`s.

**Reply by bielikb** [reply]
Thanks for the answer.

Yeah, you're right. The store itself can be tested (unit tests) in isolation with ease. 
After revisiting my question here, I think you answered it. This is out of the scope of TCA itself. 
I'll keep the declaration of the main store as it is.

--

What I'm currently missing in Xcode is something similar to `PreviewProvider` (maybe `SceneProvider` ?) for previewing the `Scenes` in the canvas. 
In that case I could leverage the property dependency injection to provide the main `store` when previewing the main app (eg turning the `store` property from `private let` to `var`).

Anyway. thanks!

Best,
Boris

**Reply by lukeredpath** [reply]
I don't think there's a whole lot to be gained from trying to test your App struct. It's effectively an integration point so you could write some kind of integrated test to make sure it's wired up properly - a UI test or a snapshot test might be sufficient here. 

You should be able to inject the store from your App into your main scene though and I think you can use SwiftUI previews from this level down.
